Add convenience methods for `BusinessOpeningHours`

gtkacz opened the pull request:
<!--
Hey! You're PRing? Cool!
Please be sure to check out our contribution guide (https://github.com/python-telegram-bot/python-telegram-bot/blob/master/.github/CONTRIBUTING.rst).
Especially, please have a look at the check list for PRs (https://github.com/python-telegram-bot/python-telegram-bot/blob/master/.github/CONTRIBUTING.rst#check-list-for-prs). Feel free to copy (parts of) the checklist to the PR description to remind you or the maintainers of open points or if you have questions on anything.
-->

Closes #4194

Bibo-Joshi commented:
Hey, thank you vermuch for your enthusiasm! As you've probably seen in #4194, we were going to consider that issue only after dropping py3.8 in ~november. TBH I don't think that we'd be willing to add an additional py3.8-dependency for this …
I see that your PR is marked as WIP in any case, so there's no hurry and you're ofc very much welcome to prepare your PR so that it can be considered directly after we drop py3.8 :) Still, I wanted to give you a heads up.

gtkacz commented:
@Bibo-Joshi hey there! Removed the Python<3.9 support and cleaned up the tests. The hooks that are malfunctioning are from code that isn't mine, but in the same module. Should I fix them?

harshil21 commented:
> @Bibo-Joshi hey there! Removed the Python<3.9 support and cleaned up the tests. The hooks that are malfunctioning are from code that isn't mine, but in the same module. Should I fix them?

Yes, most of them are line too long errors. Just run `black` and it should be fine. `pre-commit` must pass for every change.

gtkacz commented:
> > @Bibo-Joshi hey there! Removed the Python<3.9 support and cleaned up the tests. The hooks that are malfunctioning are from code that isn't mine, but in the same module. Should I fix them?
> 
> Yes, most of them are line too long errors. Just run `black` and it should be fine. `pre-commit` must pass for every change.

Actually, there's some time out errors on some auto test runs, how do I solve those?

harshil21 commented:
> Actually, there's some time out errors on some auto test runs, how do I solve those?

don't worry about them, they're just flaky.

gtkacz commented:
> > Actually, there's some time out errors on some auto test runs, how do I solve those?
> 
> don't worry about them, they're just flaky. However, `test-documentation-build` must pass too.

Trying to fix the errors, found this: https://stackoverflow.com/questions/11417221/sphinx-autodoc-gives-warning-pyclass-reference-target-not-found-type-warning

The warnings were:

```
D:\Codes\python-telegram-bot\telegram\_business.py:docstring of telegram._business.BusinessOpeningHours.get_opening_hours_for_day:1: WARNING: py:class reference target not found: date
D:\Codes\python-telegram-bot\telegram\_business.py:docstring of telegram._business.BusinessOpeningHours.get_opening_hours_for_day:1: WARNING: py:class reference target not found: optional
D:\Codes\python-telegram-bot\telegram\_business.py:docstring of telegram._business.BusinessOpeningHours.get_opening_hours_for_day:1: WARNING: py:class reference target not found: datetime
D:\Codes\python-telegram-bot\telegram\_business.py:docstring of telegram._business.BusinessOpeningHours.get_opening_hours_for_day:1: WARNING: py:class reference target not found: datetime
D:\Codes\python-telegram-bot\telegram\_business.py:docstring of telegram._business.BusinessOpeningHours.is_open:1: WARNING: py:class reference target not found: datetime
```

harshil21 commented on tests/test_business.py:
```diff
@@ -410,3 +425,51 @@ def test_equality(self):
+
+        opening_hours_for_day = business_hours.get_opening_hours_for_day(target_date)
+
+        assert opening_hours_for_day[0][0] == expected_opening_time
+        assert opening_hours_for_day[0][1] == expected_closing_time
```
please also add a negative assertion, i.e. make sure the opening hours of Tuesday are NOT returned

harshil21 commented on tests/test_business.py:
```diff
@@ -410,3 +425,51 @@ def test_equality(self):
         assert boh1 != boh3
         assert hash(boh1) != hash(boh3)
+
+    def test_get_opening_hours_for_day_success_without_timezone(
```
```suggestion
    def test_get_opening_hours_for_day_without_timezone(
```

harshil21 commented on tests/test_business.py:
```diff
@@ -410,3 +425,51 @@ def test_equality(self):
+            target_date, tzinfo=test_tz
+        )
+
+        assert opening_hours_for_day[0][0] == expected_opening_time
+        assert opening_hours_for_day[0][1] == expected_closing_time
```
same here - negative assertion

harshil21 reviewed (changes requested):
good work!

> The warnings were:

See my full review below, you can find the fix in it.

gtkacz commented:
@harshil21 finally got the time to fix all CR requests! Still can't get some pre-commit hooks to pass though, black doesn't formatt anything and ruff formats the entire codebase lol

Bibo-Joshi commented:
Hi. Black doesn't add line breaks in docstrings. Please manually add those. To sort the import blocks, please run `ruff check . --fix` in the ptb root.

gtkacz commented:
@Bibo-Joshi getting this error

```
D:\Codes\python-telegram-bot>ruff check . --fix
ruff failed
  Cause: Failed to parse D:\Codes\python-telegram-bot\pyproject.toml
  Cause: TOML parse error at line 125, column 1
    |
125 | [tool.ruff.lint]
    | ^^^^^^^^^^^^^^^^^
Unknown rule selector: `DOC`
```

gtkacz commented:
I can't get both isort/Black and Ruff to pass lol

Bibo-Joshi commented:
Please double check that you have installed the same versions of ruff, black and isort as listed in the pre-commit.yaml file

gtkacz commented:
@Bibo-Joshi I have checked, Ruff and isort don't agree on sorting statements. When I run one, the other fails.

Bibo-Joshi commented:
I can look into this once we drop 3.8. Maybe it's also related to which python versions run on

gtkacz commented:
@Bibo-Joshi the issue is that Ruff is recognizing `zoneinfo` as a non-stdlib package for some reason and placing it after the `datetime` imports. isort is the correct import format as per PEP-8.

gtkacz commented:
Finally fixed with the help of Ruff's [AlexWaygood](https://github.com/AlexWaygood) via a toml file configuration. Check out the full discussion at https://github.com/astral-sh/ruff/issues/13354 if y'all want.

harshil21 reviewed (approved):
lgtm!

Bibo-Joshi commented on pyproject.toml:
```diff
@@ -144,6 +144,10 @@ select = ["E", "F", "I", "PL", "UP", "RUF", "PTH", "C4", "B", "PIE", "SIM", "RET
 [tool.ruff.lint.pydocstyle]
 convention = "google"
 
+# CAN BE REMOVED AFTER SUPPORT FOR PYTHON 3.8 IS DROPPED
+[tool.ruff.lint.isort]
+extra-standard-library = ["zoneinfo"]
+
```
todo

Bibo-Joshi commented on telegram/_business.py:
```diff
@@ -453,3 +455,69 @@ def de_json(
+                A tuple of tuples containing the opening and closing times for the day.
+        """
+        if tzinfo is None:
+            tzinfo = zoneinfo.ZoneInfo(self.time_zone_name)
```
https://docs.python.org/3.11/library/zoneinfo.html#data-sources

example error:

```python
zoneinfo._common.ZoneInfoNotFoundError: 'No time zone found with key Europe/Berlin'
```

IMO we should handle this a bit better and tell the user that they should install `tzdata` in that case. I recommend  using try-except and reraising a `ZoneInfoNotFoundError` with a more descriptive error message. It will be good to provide a small wrapper function for that in `telegram/_utils/datetime.py` so that it can be reused (see also line 509)

Bibo-Joshi commented on telegram/_business.py:
```diff
@@ -453,3 +455,69 @@ def de_json(
+        output = []
+
+        for interval in self.opening_hours:
+            opening_time = datetime.combine(target_date, time(*interval.opening_time[1:]), tzinfo)
+            closing_time = datetime.combine(target_date, time(*interval.closing_time[1:]), tzinfo)
```
IISC this blindly assumes that the `interval` is valid for the `target_data`. But the business might be closed on that date. This doesn't look right to me. If I'm wrong, kindly explain why this works

Bibo-Joshi commented on tests/test_business.py:
```diff
@@ -410,3 +416,83 @@ def test_equality(self):
+
+        assert opening_hours_for_day[0][0].date() != target_date_plus_1
+        assert opening_hours_for_day[0][1].date() != target_date_plus_1
+        assert opening_hours_for_day[1][0].date() != target_date_plus_1
+        assert opening_hours_for_day[1][1].date() != target_date_plus_1
```
`opening_hours_for_day` must have only one entry - you're requesting tha opening slots for monday and there is only one `BusinessOpeningHoursInterval` defined for monday

Bibo-Joshi reviewed (changes requested):
Hi. As #4368 is getting closer to merge, I had a first rough look at the changes. There are two larger concerns that I have with them.

Bibo-Joshi commented:
@gtkacz are you still interesting in finishing up this pr?

gtkacz commented:
@Bibo-Joshi interested I am, just haven't had the time lately to do so lol, will try to get to it this weekend
